EIPS/eip-lock_sload.md

```diff
@@ -0,0 +1,109 @@
+
+That's exactly what happened to the DAO and can happen to anyone else in the future: 2 users (frames) are operating on the same record (balance of tokens for example) and they are doing the same operation more times that it was supposed to be done.
+
+Most traditional databases do not lock records by default. Which had its consequinces off-chain either: 
+
+http://hackingdistributed.com/2014/04/06/another-one-bites-the-dust-flexcoin/
+
+These bugs are literally everywhere. __Except, none of them yielded $150M on an unreversible public ledger.__ So unlike traditional databases, Ethereum like no one else should take this problem seriously.
+
+Why does it happen and is not prevented on the database level? Because financial stuff is a tiny part of usage for them (__even for financial applications, fiat race condition is not a big deal as it can be reversed with a little hassle__). If they locked the tables/records _by default_, they would be way too slow _by default_ albeit much more secure against isolation vulnerabilities.
+
+That's why __they__ don't do that, but __we should__. That's why a lot of traditional apps can be hacked by simply sending 10 parallel requests instead of one.
+
+Re-entrancy is no different from a race condition: the same contract is invoked in the same frame stack, while they both can work on the same record from the table (not exactly in parallel, but frame 3 in the middle of frame 1). 
+
+And Ethereum's key-value contract storage, unlike traditional databases, _is all about finance_ and highly sensitive balances/amounts. On top of that the extra locking performance overhead doesn't cost anything for a blockchain and its use cases. 
+
+So since the storage is __a database that operates only with sensitive values__ - that's where we need isolation the most. Let's introduce __pessimistic locking by default__ and redefine the SLOAD opcode. That does not break a single valid usage of a contract, yet prevents DAO-like attacks including the DAO itself (__which will be considered a technical fork over a political one from now on__ if this EIP is accepted).
+
+## Specification
+
+Need your advices on this one. I'm sure the idea is outright obvious. The goal is to lock a key that was SLOAD-ed, i.e. turn `select value from Storage0xA where key="Key0xA"` into `select value from Storage0xA where key="Key0xA" SELECT FOR UPDATE`.
+
+Entire Storage0xA locking is probably an overkill. That would kill a few use-cases when the contract is called again in the same frame stack yet operates with different storage keys - no race condition bug there. Instead we should add a global lock array. After each frame termination the locks ackquired in this particular frame must be released.
+
+
+```
+def SLOAD(account, key)
+  lock_id = account+key
+  if $locked.includes(lock_id)
+    throw;
+  else
+    $locked.push(account)
+    old_SLOAD(account, key)
+  end
+end
+
```

Comment:
What about `SSTORE`?  I guess `SSTORE` needs to be modified.  If that's not the case, that should be said explicitly I think.

Reply:
I'm busy with something else and in this threat there were better suggestions by @Arachnid  so this EIP isn't probably best way. I cannot keep it up to date - feel free to use create new one.

As of sstore, the safe way would be to throw error if the key is still locked, but most of the time you load before writing, so it would throw anyway.

Reply:
@homakov thank you for the response.  I marked this PR with `looking-for-champion`.